cleanlab/rank.py

```diff
@@ -556,52 +554,59 @@ def get_confidence_weighted_entropy_for_each_label(
+    """Returns the KNN distance outlier score for each datapoint.
 
-    This is a function to compute OOD scores where higher scores indicate the datapoint is more likely to be OOD.
+    This is a function to compute outlier scores where higher scores indicate the datapoint is more likely to be an outlier.
 
     Parameters
     ----------
     features : np.array
-      Feature matrix of shape (N, M), where N is the number of datapoints and M is the number of features.
+      Feature matrix of shape ``(N, M)``, where N is the number of datapoints and M is the number of features.
       This is the "query set" of features for each datapoint which are used for nearest neighbor search.
 
-    nbrs : sklearn.neighbors.NearestNeighbors
-      Instantiated NearestNeighbors class object that's been fitted on a dataset in the same feature space.
+    knn : sklearn.neighbors.NearestNeighbors
+      Instantiated ``NearestNeighbors`` class object that's been fitted on a dataset in the same feature space.
       Note that the distance metric and n_neighbors is specified when instantiating this class.
+      You can also pass in a subclass of ``sklearn.neighbors.NearestNeighbors`` which allows you to use faster
+      approximate neighbor libraries as long as you wrap them behind the same sklearn API.
+      If ``knn = None``, then by default ``knn = sklearn.neighbors.NearestNeighbors(n_neighbors=k, metric="cosine").fit(features)``
       See: https://scikit-learn.org/stable/modules/neighbors.html
 
-    k : int, default=None
-      Number of neighbors to use when calculating average distance to neighbors.
-      This value k needs to be less than or equal to max_k which is the n_neighbors used when fitting instantiated NearestNeighbors class object.
-      If k=None, then by default k=min(10, max_k) is used where max_k is extracted from the given nbrs.
+    k : int, default=10
+      Number of neighbors to use when calculating outlier score (average distance to neighbors). If an existing ``knn``
+      object is provided, you can still specify that outlier score should use a different value of ``k`` than originally
+      used in the ``knn``, as long as your specified value of ``k`` is smaller than the one originally used.
+      If k is larger, then by default ``k = n_neighbors`` where ``n_neighbors`` is defined in the ``knn``.
+      If k is not provided, then by default ``k = 10``.
 
     Returns
     -------
-    avg_nbrs_distances : np.array
-      Average distance to k-nearest neighbors for each datapoint which is used as a score for OOD detection.
+    avg_knn_distances : np.array
+      Average distance to k-nearest neighbors for each datapoint which is used as a score for outlier detection.
     """
+    # if knn is not provided, then use default KNN as estimator
+    if knn is None:
+        knn = NearestNeighbors(n_neighbors=k, metric="cosine").fit(features)
```

Comment:
after this (while still inside of the `if` clause), we should do:

```
features = None
```
so that it'll be None in the later call to: `knn.kneighbors(features)`.

When kneighbors gets None as its input it returns the neighbors for the training data:
https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors.kneighbors

which is what we want to not waste computation (and to get the number of neighbors correct, otherwise the datapoint itself was being counted as a neighbor of itself)

Reply:
Wouldn't the same problem occur in the original call:
```
knn = NearestNeighbors(n_neighbors=k, metric="cosine").fit(features)
outlier_score = get_outlier_scores(features, knn, k)
```

Reply:
Hacky way to enforce that would be
```
if np.array_equal(knn.__dict__['_fit_X'], features):
        features = None
```

Reply:
Good point!  I'd just clarify the following things at the end of the `knn` arg docstring:

If you specify ` ``knn`` ` here and wish to find outliers in the same data you already passed into ` ``knn.fit(features)`` `, then you should simply specify ` ``features = None`` ` here.  This requires that your ` ``knn.kneighbors(None)`` `
returns the distances to the datapoints it was ` ``fit()`` ` on.  Not recommended but if you specify ` ``features`` `  here again as the same features previously used in ` ``knn.fit()`` `, then each example will be also counted as one of its own neighbors and the average KNN distances will be skewed downward (although their relative ordering will be preserved).

Reply:
I'd want to avoid your route in order to not impose more requirements on what the user's custom `knn` class has to satisfy (we want to be compatible with as many approximate neighbor libraries as possible here)

Reply:
@jwmueller is this hacky way ok, that is my most recent push.

Reply:
Let's not do that for the reasons mentioned above. We really don't want the user's custom knn class to have to conform a lot of specifics, and I don't think it's a big deal if they happened to pass in features twice. So let's just clarify this issue in the docstring as I mentioned above and not do anything about this issue in the source code.

Reply:
Docstring for ``knn`` looked SO long if I included that entire (great!) explanation. Shortened it a little, let me know if that still conveys same information.

Reply:
shortening my overly long statements is always a good idea :) 
New docstring looks great!